sstables/storage.cc

```diff
@@ -36,10 +36,57 @@
 
 namespace sstables {
 
+class dir_wrapper final {
+    std::filesystem::path _pathname;
+    file _file;
+    bool _file_stale;
+
+public:
+    explicit dir_wrapper(std::filesystem::path pathname) : _pathname(std::move(pathname)) {};
+    dir_wrapper(const dir_wrapper&) = delete;
+    dir_wrapper& operator=(const dir_wrapper&) = delete;
+    dir_wrapper(dir_wrapper&&) = delete;
+    dir_wrapper& operator=(dir_wrapper&&) = delete;
+    ~dir_wrapper() = default;
+
+    const std::filesystem::path::string_type& native() const noexcept {
+        return _pathname.native();
+    }
+
+    operator std::filesystem::path() const noexcept {
+        return _pathname;
+    }
```

Comment:
A method returning reference to the path member would be clearer than the implicit type-conversion operator.

Reply:
I added the conversion operator in order to minimize the impact for the rest of the code.

With the following change:
```
diff --git a/sstables/storage.cc b/sstables/storage.cc
index 5ee463c00ef9..dbc1be27ea78 100644
--- a/sstables/storage.cc
+++ b/sstables/storage.cc
@@ -53,7 +53,7 @@ class dir_wrapper final {
         return _pathname.native();
     }
 
-    operator std::filesystem::path() const noexcept {
+    const std::filesystem::path& as_path() const noexcept {
         return _pathname;
     }
 
```

the impact is as follows:

```
@@ -166,7 +166,7 @@ future<> filesystem_storage::rename_new_file(const sstable& sst, sstring from_na
 future<file> filesystem_storage::open_component(const sstable& sst, component_type type, open_flags flags, file_open_options options, bool check_integrity) {
     auto create_flags = open_flags::create | open_flags::exclusive;
     auto readonly = (flags & create_flags) != create_flags;
-    auto tgt_dir = !readonly && _temp_dir ? *_temp_dir : _dir;
+    auto tgt_dir = !readonly && _temp_dir ? *_temp_dir : _dir.as_path();
     auto name = tgt_dir / sst.component_basename(type);
 
     auto f = open_sstable_component_file_non_checked(name.native(), flags, options, check_integrity);
@@ -232,7 +232,7 @@ future<> filesystem_storage::touch_temp_dir(const sstable& sst) {
     if (_temp_dir) {
         co_return;
     }
-    auto tmp = _dir / fmt::format("{}{}", sst._generation, tempdir_extension);
+    auto tmp = _dir.as_path() / fmt::format("{}{}", sst._generation, tempdir_extension);
     sstlog.debug("Touching temp_dir={}", tmp);
     co_await sst.sstable_touch_directory_io_check(tmp);
     _temp_dir = std::move(tmp);
@@ -404,7 +404,7 @@ future<> filesystem_storage::snapshot(const sstable& sst, sstring dir, absolute_
     if (abs) {
         snapshot_dir = dir;
     } else {
-        snapshot_dir = _dir / dir;
+        snapshot_dir = _dir.as_path() / dir;
     }
     co_await sst.sstable_touch_directory_io_check(snapshot_dir);
     co_await create_links_common(sst, snapshot_dir, std::move(gen));
@@ -433,7 +433,7 @@ future<> filesystem_storage::move(const sstable& sst, sstring new_dir, generatio
 
 future<> filesystem_storage::change_state(const sstable& sst, sstable_state state, generation_type new_generation, delayed_commit_changes* delay_commit) {
     auto to = state_to_dir(state);
-    std::filesystem::path path = _dir;
+    std::filesystem::path path = _dir.as_path();
     auto current = path.filename().native();
 
     // Moving between states means moving between basedir/state subdirectories.
```

Reply:
Doing this in v5.

Reply:
These changes are listed in the v5 changelog; resolving this thread.